Reduce unsafe member variables in WebCore scrolling code

annevk opened the pull request:
#### 8e82e43489dafd2f1868b62442306fcfc6d0c401
<pre>
Reduce unsafe member variables in WebCore scrolling code
<a href="https://bugs.webkit.org/show_bug.cgi?id=304762">https://bugs.webkit.org/show_bug.cgi?id=304762</a>

Reviewed by Charlie Wolfe.

We use RefPtr for the members of the NodeLayers struct as it&apos;s
stack-allocated.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

Canonical link: <a href="https://commits.webkit.org/305024@main">https://commits.webkit.org/305024@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86a0640a0035c52a7beb6247a0ef7305d49604c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137203 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9563 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48490 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144952 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90176 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c17dc18f-8f3f-4344-8204-21cbfbef41f2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139075 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10266 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9690 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104924 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/433caca3-4e6a-4bf3-8499-361fb9a00ef1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140148 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7567 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122955 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85765 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/71f2f45d-4f61-4065-b318-52cd06500a09) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7204 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4915 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5539 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116560 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41118 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147710 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9246 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41680 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113284 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9264 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113614 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7123 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119210 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63683 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21141 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9295 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37264 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9020 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72860 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9235 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9087 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/scrolling/ScrollingTreeGestureState.h:
```diff
@@ -29,6 +29,7 @@
 
 #include <WebCore/ScrollTypes.h>
 #include <wtf/Markable.h>
+#include <wtf/WeakPtr.h>
```
Wrong include for `ThreadSafeWeakPtr`. Should include `<wtf/ThreadSafeWeakPtr.h>`

cdumez commented on Source/WebCore/rendering/LayerAncestorClippingStack.cpp:
```diff
@@ -115,7 +115,17 @@ void LayerAncestorClippingStack::updateScrollingNodeLayers(ScrollingCoordinator&
-        scrollingCoordinator.setNodeLayers(*entry.overflowScrollProxyNodeID, { entry.scrollingLayer.get() });
+        scrollingCoordinator.setNodeLayers(*entry.overflowScrollProxyNodeID, {
+            entry.scrollingLayer.get(),
+            nullptr,
```
I feel like it was better to keep the `{ nullptr }` in the struct, so we don't have to explicitly specify them here at the call site.

charliewolfe commented on Source/WebCore/page/scrolling/ScrollingCoordinator.h:
```diff
@@ -156,15 +156,15 @@ class ScrollingCoordinator : public ThreadSafeRefCountedAndCanMakeThreadSafeWeak
+        WeakPtr<GraphicsLayer> rootContentsLayer { nullptr };
+        WeakPtr<GraphicsLayer> horizontalScrollbarLayer { nullptr };
+        WeakPtr<GraphicsLayer> verticalScrollbarLayer { nullptr };
+        WeakPtr<GraphicsLayer> viewportAnchorLayer { nullptr };
```
`NodeLayers` is always stack-allocated. We should be able to use `RefPtr` here to prevent these layers from being destroyed without any lifetime issues. Then all the uses of this in `AsyncScrollingCoordinator.cpp` will be safe.

Also the `{ nullptr }` default initialization is not needed.

annevk replied:
Without the initialization, implicitly allocating the struct will fail as variable number of arguments are used. At least with WeakPtr that’s the case.

cdumez replied:
Right, `{ nullptr }` is useful given that this is a POD struct and given the way it is initialized. Charlie has a point though that this is always stack-allocated and thus we should use RefPtr instead.

charliewolfe commented on Source/WebCore/page/scrolling/ScrollingTreeGestureState.cpp:
```diff
@@ -93,11 +93,11 @@ void ScrollingTreeGestureState::nodeDidHandleEvent(ScrollingNodeID nodeID, const
     case PlatformWheelEventPhase::Ended:
         if (m_activeNodeID)
-            m_scrollingTree.handleWheelEventPhase(*m_activeNodeID, event.momentumPhase());
+            m_scrollingTree.get()->handleWheelEventPhase(*m_activeNodeID, event.momentumPhase());
```
We should store this in a `Ref` before using it.

annevk replied:
Why?

charliewolfe replied:
`m_scrollingTree` is a weak reference, so it won't prevent the object from being destroyed. Don't we need a strong reference on the stack since `handleWheelEventPhase` and others are non-trivial functions?

annevk replied:
get() returns a RefPtr because it’s a ThreadSafeWeakPtr.

charliewolfe replied:
I see, I missed that

charliewolfe commented on Source/WebCore/page/scrolling/ScrollingTreeGestureState.h:
```diff
@@ -50,7 +51,7 @@ class ScrollingTreeGestureState {
     void clearAllNodes();
 
-    ScrollingTree& m_scrollingTree;
+    ThreadSafeWeakPtr<ScrollingTree> m_scrollingTree; // Cannot be null.
```
Eventually we should make a `ThreadSafeWeakRef`

annevk replied:
#55961 🎉

annevk committed:
Reduce unsafe member variables in WebCore scrolling code
https://bugs.webkit.org/show_bug.cgi?id=304762

Reviewed by Charlie Wolfe.

We use RefPtr for the members of the NodeLayers struct as it's
stack-allocated.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

Canonical link: https://commits.webkit.org/305024@main
